snuba/pipeline/query_pipeline.py

```diff
@@ -76,3 +81,51 @@ def build_planner(
         self, query: Query, settings: QuerySettings
     ) -> QueryPlanner[ClickhouseQueryPlan]:
         raise NotImplementedError
+
+
+class QueryPipelineStage(Generic[Tin, Tout]):
+    """
+    This class represents a single stage in the snuba query execution pipeline.
+    The purpose of this class is to provide an organized and transparent interface to
+    execute specific processing steps on the query with clearly defined inputs and outputs.
+    These stages are designed to be composed and/or swapped among each other to form a
+    a flexible query pipeline.
+
+    Some examples of a query pipeline stage may include:
+    * Execute all entity query processors defined on the entity yaml
+    * Apply query transformation from logical representation to Clickhouse representation
+    * Execute all storage processors defined on the storage yaml
+    * Run query execution
+    * Query reporting
+
+    To create a Query Pipeline Stage, the main components to specify are:
+    an input type, an output type, and a execution function which returns the output wrapped with QueryPipelineResult.
+    ==============================================
+        >>> class MyQueryPipelineStage(QueryPipelineStage[LogicalQuery, LogicalQuery]):
+        >>>    def _execute(self, input: QueryPipelineResult[LogicalQuery]) -> QueryPipelineResult[LogicalQuery]:
+        >>>         try:
+        >>>             result = my_complex_processing_function(input.data)
+        >>>             return QueryPipelineResult(result, None)
+        >>>         except Exception as e:
+        >>>             return QueryPipelineResult(None, e)
+    """
+
+    @abstractmethod
+    def _execute(self, input: QueryPipelineResult[Tin]) -> QueryPipelineResult[Tout]:
```

Comment:
It seems to me that `execute` should be the only interface method we define since from a top down view its the only thing we need a stage to do, execute input to give output. If the user wants to define a private helper function for execute they can do it or not, it doesnt matter.

Reply:
This is indeed the only method a user would need to define. That method is wrapped by default to allow expansion of the capabilities of the interface

Reply:
Ohh I see, that's cool.